AirCasting/MicrophoneSession/MicrophoneManager.swift

```diff
@@ -21,86 +18,110 @@ class MicrophoneManager: ObservableObject {
+        if audioSession.recordPermission == .granted {
+            addInterruptionsObserver()
+            
+            self.session = session
+            createDBStream()
+            locationProvider.requestLocation()
+            
+            let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
+            
+            let recordSettings: [String: Any] = [
+                AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
+                AVSampleRateKey: 44100.0,
+                AVNumberOfChannelsKey: 1,
+                AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
+            ]
+            
+            do {
+                try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
+                try audioSession.setActive(true)
+                try recorder = AVAudioRecorder(url:url, settings: recordSettings)
+                
+            } catch {
+                return
             }
-        }
-        
-        self.session = session
-        createDBStream()
-        locationProvider.requestLocation()
-        
-        let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
-        
-        let recordSettings: [String: Any] = [
-            AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
-            AVSampleRateKey: 44100.0,
-            AVNumberOfChannelsKey: 1,
-            AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
-        ]
-        
-        do {
-            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
-            try audioSession.setActive(true)
-            try recorder = AVAudioRecorder(url:url, settings: recordSettings)
             
-        } catch {
-            return
+            recorder.isMeteringEnabled = true
+            recorder.record()
+            isRecording = true
+            
+            levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
         }
-        
-        recorder.isMeteringEnabled = true
-        recorder.record()
-        isRecording = true
-        
-        levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
     }
     
     @objc func getMeasurement() {
+        if !recorder.isRecording {
+            recorder.record()
+        }
         recorder.updateMeters()
-        
         let level = recorder.averagePower(forChannel: 0)
+        createMeasurement(value: level)
+    }
+    
+    func stopRecording() {
+        levelTimer.invalidate()
+        isRecording = false
+        recorder.stop()
+        recorder = nil
+        session?.status = Int16(SessionStatus.FINISHED.rawValue)
+        do {
+            try context.save()
+        } catch {
+            print("Error for saving session when stopping mic recording: \(error)")
+        }
+    }
+    
+    @objc func handleInterruption(notification: Notification) {
```

Comment:
please make it private. except the Notification Center here there should be no direct calls to this method form outside of the class.